File: Pi_analysis.m
%Analysis of number density in Pi digits

%Created by Yokesh
%Email: yzs0074@auburn.edu

clear;clc;
close all;

%Import the Pi data text file
%Read the entire .txt file as a string
%Beware you should have enough to memory to handle, when to try >1e9 digits
data = fileread("pi_dec_1m.txt");
data = data(3:end); %To elminiate 3. at the start

tracker = zeros(10,1); %Tracker counts #digit
n = length(data);

for i = 1:n

    %Rips the string and converts into a double data type
    %Then it is converted to int data type using 'floor'
    cc = floor(str2double(data(i)));  %I know, but this is the faster way in Matlab
    tracker(cc+1) = tracker(cc+1) + 1; %Update the tracker
end

percent = zeros(10,1); %Percent calculates percentage of occurence of each #digit
for i = 1:10
    percent(i) = tracker(i)/n*100;
end

%Plot
figure(1);
plot(0:9, percent, 'k*');
title("Pi digits - analysis");
xlabel("#digits");
% axis([-1 10 9.9 10.2]); (optional)
ylabel("Percentage");

File: Pi_analysis.py
#Analysis of Pi digits

#Created by Yokesh
#Email: yzs0074@auburn.edu

import numpy as np
import matplotlib.pyplot as plt
import time

#Input the .txt file from the link in the description
f = open("pi_dec_1m.txt", "r");

#Read the entire .txt file as a string
#Beware you should have enough to memory to handle, when to try >1e9 digits
data = f.read()

#Initialize
n = len(data)
tracker = np.zeros([10,1], dtype = np.int_)   #Tracks the count of each #digit
percent = np.zeros([10,1], dtype = np.float_) # To find percent of each #digit
x = np.zeros([10,1], dtype = np.int_)         # Variable to plot (optional)

start_time = time.time()   #Keep track of time (optional)

#Remember the loop runs from third character to eliminate 3. at the start
for i in range(2, n):
    tracker[int(data[i])] += 1
    
print("--- %.8f seconds ---" % (time.time() - start_time))  #Prints the time elapsed (optional)
f.close()
 
#Loop to calculate the percent of each #digit
for i in range(0, len(tracker)):
    percent[i] = tracker[i]/(n-2)*100;
    x[i] = i
    
#Plotting
plt.plot(x, percent, "k*")
plt.xlabel("#digits")
plt.ylabel("Percentage of occurance")
plt.title("Occurance of #digits in first 1 million digits of Pi")

#To display the values of each data point
#This section yet to updated with beautiful looking data points and x-axis labels
for a,b in zip(x, percent): 
     plt.text(a, b, str(b))
    
plt.show()
    
    
